### PR TITLE
[Consent management] Add Segment consent changed event

### PR DESCRIPTION
**Pull request by silesky**
- Add `registerOnConsentChanged` setting.
- For OneTrust, dispatch a segment consent changed event whenever OneTrust consent change.
- Add ability to disable consent changed events.
- update README.

Other misc goodies!

### CODE REVIEW

File: packages/consent/consent-tools/src/domain/create-wrapper.ts [hunk 1]
```diff
@@ -118,6 +125,19 @@ export const createWrapper: CreateWrapper = (createWrapperOptions) => {
...
+        registerOnConsentChanged((categories) => {
+          try {
+            validateCategories(categories)
+            sendConsentChangedEvent(analytics, categories)
+          } catch (err) {
+            // Not sure if there's a better way to handle this, but this makes testing a bit easier.
+            console.error(err)
+          }
+        })
+      }
```
Comment (oscb) [inline packages/consent/consent-tools/src/domain/create-wrapper.ts:139]: nit: you could do `registerOnConsentChanged?.(`

File: packages/consent/consent-tools/src/domain/validation/common-validators.ts [hunk 2]
```diff
@@ -11,7 +11,7 @@ export function assertIsFunction(
 
 export function assertIsObject(
   val: unknown,
-  variableName: string
+  variableName = 'value'
```
Comment (oscb) [inline packages/consent/consent-tools/src/domain/validation/common-validators.ts:14]: shouldn't it keep its typedef to `string`? won't TS restrict this to `value` only?

File: packages/consent/consent-wrapper-onetrust/src/domain/wrapper.ts [hunk 3]
```diff
@@ -1,25 +1,30 @@
...
   integrationCategoryMappings?: CreateWrapperSettings['integrationCategoryMappings']
+  disableConsentChangedEvent?: boolean
 }
 
+/**
+ *
+ * @param analyticsInstance - An analytics instance. Either `window.analytics`, or the instance returned by `new AnalyticsBrowser()` or `AnalyticsBrowser.load({...})`
+ * @param settings - Optional settings for configuring your OneTrust wrapper
+ */
 export const oneTrust = (
```
Comment (oscb) [inline packages/consent/consent-wrapper-onetrust/src/domain/wrapper.ts:24]: nit: should this be named `withOneTrust` instead? To make sure people understand it's wrapping the `analytics` object?

File: packages/consent/consent-tools/src/types/settings.ts [hunk 4]
```diff
@@ -27,6 +27,36 @@ export interface CreateWrapperSettings {
    **/
   getCategories: () => Categories | Promise<Categories>
 
+  /**
+   * Programatically send a "Segment Consent Preference" event to Segment when consent preferences change.
+   * An event will be sent to Segment whenever this function is called.
```
Comment (oscb) [inline packages/consent/consent-tools/src/types/settings.ts:32]: I would specify that this is a "Function to register a listener for consent changes to programatically send ..."

**Review by oscb — APPROVED** [review]
Left a few nits, nothing is a blocker imo

File: packages/consent/consent-tools/src/domain/validation/common-validators.ts (hunk 2 above)
Comment (silesky) [reply packages/consent/consent-tools/src/domain/validation/common-validators.ts:14]: Nope, it keeps the string type (not a const)

File: packages/consent/consent-wrapper-onetrust/src/domain/wrapper.ts (hunk 3 above)
Comment (silesky) [reply packages/consent/consent-wrapper-onetrust/src/domain/wrapper.ts:24]: I've been weighing that change as well, I'm glad you chimed in

**Pull request merged**